pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+			// User asked for a PV that does not exist.
+			// OBSERVATION: pvc is "Pending"
+			// Retry later.
+			glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume %q requested and not found, will try again next time", claimToClaimKey(claim), claim.Spec.VolumeName)
+			if _, err = ctrl.updateClaimPhase(claim, api.ClaimPending); err != nil {
+				return err
+			}
+			return nil
+		} else {
+			volume, ok := obj.(*api.PersistentVolume)
+			if !ok {
+				return fmt.Errorf("Cannot convert object from volume cache to volume %q!?", claim.Spec.VolumeName)
+			}
+			glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume %q requested and found: %s", claimToClaimKey(claim), claim.Spec.VolumeName, ctrl.getVolumeStatusForLogging(volume))
+			if volume.Spec.ClaimRef == nil {
+				// User asked for a PV that is not claimed
+				// OBSERVATION: pvc is "Pending", pv is "Available"
+				glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume is unbound, binding", claimToClaimKey(claim))
+				if err = ctrl.bind(volume, claim); err != nil {
+					// On any error saving the volume or the claim, subsequent
+					// syncClaim will finish the binding.
+					return err
+				}
+				// OBSERVATION: pvc is "Bound", pv is "Bound"
+				return nil
+			} else if isVolumeBoundToClaim(volume, claim) {
+				// User asked for a PV that is claimed by this PVC
+				// OBSERVATION: pvc is "Pending", pv is "Bound"
+				glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume already bound, finishing the binding", claimToClaimKey(claim))
+
+				// Finish the volume binding by adding claim UID.
+				if err = ctrl.bind(volume, claim); err != nil {
+					return err
+				}
+				// OBSERVATION: pvc is "Bound", pv is "Bound"
+				return nil
+			} else {
+				// User asked for a PV that is claimed by someone else
+				// OBSERVATION: pvc is "Pending", pv is "Bound"
+				if !ctrl.hasAnnotation(claim.ObjectMeta, annBoundByController) {
+					glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume already bound to different claim by user, will retry later", claimToClaimKey(claim))
+					// User asked for a specific PV, retry later
+					if _, err = ctrl.updateClaimPhase(claim, api.ClaimPending); err != nil {
+						return err
+					}
+					return nil
+				} else {
+					// This should never happen because someone had to remove
+					// annWasEverBound annotation on the claim.
+					glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume already bound to different claim %q by controller, THIS SHOULD NEVER HAPPEN", claimToClaimKey(claim), claimrefToClaimKey(volume.Spec.ClaimRef))
+					return fmt.Errorf("Invalid binding of claim %q to volume %q: volume already claimed by %q", claimToClaimKey(claim), claim.Spec.VolumeName, claimrefToClaimKey(volume.Spec.ClaimRef))
+				}
+			}
+		}
+	}
+}
+
+func (ctrl *PersistentVolumeController) syncBoundClaim(claim *api.PersistentVolumeClaim) error {
```

Comment:
do you plan to enforce preconditions here, or is it the responsibility of the caller to enforce them?